src/execution/execute.ts

```diff
@@ -525,6 +533,60 @@ export function collectFields(
   return fields;
 }
 
+function selectionSetWithFragmentArgumentsApplied(
+  exeContext: ExecutionContext,
+  fragment: FragmentDefinitionNode,
+  fragmentSpread: FragmentSpreadNode,
+): SelectionSetNode {
+  const variableDefinitions = fragment.variableDefinitions;
+  if (variableDefinitions == null) {
+    return fragment.selectionSet;
+  }
+
+  const providedArguments: Map<string, ArgumentNode> = new Map();
+  for (const arg of fragmentSpread.arguments ?? []) {
+    providedArguments.set(arg.name.value, arg);
+  }
+  const fragmentArgumentValues: Map<string, ValueNode> = new Map();
+  for (const argDef of variableDefinitions) {
+    const variableName = argDef.variable.name.value;
+    const providedArg = providedArguments.get(variableName);
+    if (providedArg != null) {
+      if (
+        providedArg.value.kind === Kind.NULL &&
+        argDef.type.kind === Kind.NON_NULL_TYPE
+      ) {
+        exeContext.errors.push(
+          new GraphQLError(
+            `Fragment argument "$${variableName}" on fragment "${fragment.name.value}" is non-null, but null was provided.`,
+            [providedArg, argDef],
+          ),
+        );
+      }
+      fragmentArgumentValues.set(variableName, providedArg.value);
+    } else if (argDef.defaultValue != null) {
+      fragmentArgumentValues.set(variableName, argDef.defaultValue);
+    } else if (argDef.type.kind === Kind.NON_NULL_TYPE) {
+      exeContext.errors.push(
+        new GraphQLError(
+          `Fragment argument "$${variableName}" on fragment "${fragment.name.value}" is required but not provided.`,
+          [fragmentSpread, argDef],
+        ),
+      );
+    }
+    // Otherwise just preserve the variable as-is: it will be treated as unset by the executor.
+  }
+
+  return visit(fragment.selectionSet, {
+    Variable(node) {
+      const replacementValue = fragmentArgumentValues.get(node.name.value);
+      if (replacementValue != null) {
+        return replacementValue;
+      }
```

Comment:
Map returns **undefined** for get("not-in-map")?! https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/get

I thought it returned `null`. I kind of **like** making the null explicit for dum-dums like me who don't know JS very well (especially as `visit` is confusing enough, and it helps me to remember that `undefined` and `null` do different things for `visit`) , but I agree it can be simplified so I will at your suggestion :)